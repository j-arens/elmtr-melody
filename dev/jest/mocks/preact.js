module.exports = {
    unmountComponent: jest.fn(),
};
