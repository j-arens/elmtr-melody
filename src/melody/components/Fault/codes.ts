export enum ErrorCodes {
    MELODY_GENERIC_FAULT = 1,
    MELODY_NO_TRACKS,
    MELODY_BAD_SOURCE,
}
