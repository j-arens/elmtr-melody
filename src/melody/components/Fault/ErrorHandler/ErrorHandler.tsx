import { State } from '@redux/type';
import { h } from 'preact';
import collectErrors from '../Collector';
import ErrorRenderer from '../ErrorRenderer';

export interface DispatchProps {
    getState: any;
}

export interface OwnProps {
    children?: JSX.Element[];
    audioInterface: HTMLAudioElement;
}

type Props = DispatchProps & OwnProps;

export default ({ getState, audioInterface, children }: Props) => {
    const errors = collectErrors(getState(), audioInterface);
    if (errors.size) {
        return <ErrorRenderer errors={errors} />;
    }
    return (
        <div class="melody-errorhandler">
            {children}
        </div>
    );
};
