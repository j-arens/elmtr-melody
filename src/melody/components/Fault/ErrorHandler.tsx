import { isEditMode } from '@melody/utils/';
import { h } from 'preact';
import BadSource from './BadSource';
import { ErrorCodes } from './codes';
import DefaultError from './DefaultError';
import FaultContainer from './FaultContainer';
import NoTracksEditor from './NoTracksEditor';
import NoTracksFront from './NoTracksFront';

interface Props {
    errors: Set<ErrorCodes>;
}

export default ({ errors }: Props) => {
    if (errors.has(ErrorCodes.MELODY_BAD_SOURCE)) {
        return (
            <FaultContainer allowReload>
                <BadSource />
            </FaultContainer>
        );
    }

    if (errors.has(ErrorCodes.MELODY_NO_TRACKS)) {
        if (isEditMode()) {
            return (
                <FaultContainer allowReload={false}>
                    <NoTracksEditor />
                </FaultContainer>
            );
        }
        return (
            <FaultContainer allowReload={false}>
                <NoTracksFront />
            </FaultContainer>
        );
    }

    return (
        <FaultContainer allowReload>
            <DefaultError />
        </FaultContainer>
    );
};
