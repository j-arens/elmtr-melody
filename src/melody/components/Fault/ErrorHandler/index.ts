import { State } from '@redux/type';
import { connect } from 'preact-redux';
import ErrorHandler, { OwnProps, StateProps } from './ErrorHandler';

const mapState = (state: State) => ({ state });

export default connect<
    StateProps,
    {},
    OwnProps
>(mapState)(ErrorHandler);
