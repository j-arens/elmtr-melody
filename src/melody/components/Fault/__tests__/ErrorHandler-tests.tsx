import { h } from 'preact';
import { render } from 'preact-render-to-string';
import { ErrorCodes } from '../codes';
import ErrorHandler from '../ErrorHandler';

describe('ErrorHandler', () => {
    it('renders BadSource component if error codes contains bad source error', () => {
        const errors = new Set([ErrorCodes.MELODY_BAD_SOURCE]);
        const component = render(<ErrorHandler errors={errors} />);
        expect(component).toMatchSnapshot();
    });

    it('renders NoTracksEditor component if there are no tracks and user is in edit mode', () => {
        global.elementor = {
            isEditMode: jest.fn().mockReturnValue(true),
        };
        const errors = new Set([ErrorCodes.MELODY_NO_TRACKS]);
        const component = render(<ErrorHandler errors={errors} />);
        expect(component).toMatchSnapshot();
    });

    it('render NoTracksFront component if there are no tracks and user is not in edit mode', () => {
        global.elementor = {
            isEditMode: jest.fn().mockReturnValue(false),
        };
        const errors = new Set([ErrorCodes.MELODY_NO_TRACKS]);
        const component = render(<ErrorHandler errors={errors} />);
        expect(component).toMatchSnapshot();
    });

    it('renders DefaultError component for unhandled errors', () => {
        const errors = new Set([9999]);
        const component = render(<ErrorHandler errors={errors} />);
        expect(component).toMatchSnapshot();
    });
});
