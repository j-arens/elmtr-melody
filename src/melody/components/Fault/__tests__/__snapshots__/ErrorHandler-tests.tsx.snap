// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorHandler renders errors if there are any 1`] = `
<div class="errorContainer">
  <p class="ErrorMsg">Something went wrong! Please try refreshing the page, otherwise contact
    the site administrator.</p>
  <button class="btn errorContainer__reloadBtn"
  role="button">Reload</button>
</div>
`;

exports[`ErrorHandler renders its children if there are no errors 1`] = `
<div class="melody-errorhandler">
  <p>lol</p>
</div>
`;
